File: de.unistuttgart.iste.sqa.mpw.framework.cpp/core/src/impl/GameViewPresenterImpl.cpp
#include "GameViewPresenterImpl.h"

#include "Stage.h"
#include "GameLog.h"
#include "Semaphore.h"
#include "CommandStack.h"
#include "GamePerformance.h"

#include "ViewModelLogEntry.h"

#include "CollectionHelpers.hpp"
#include <stdexcept>
#include <utility>

using namespace mpw;
using namespace collectionhelpers;
using namespace framework;

namespace viewmodel {

// TODO remove listeners on removement of relating objects

GameViewPresenterImpl::GameViewPresenterImpl(std::shared_ptr<mpw::MiniProgrammingWorld> miniProgrammingWorld)
    : inherited(Semaphore())
    , miniProgrammingWorld(std::move(miniProgrammingWorld)) {
    const Size& size = this->miniProgrammingWorld->getStage()->getStageSize();
    GameViewPresenter::getViewModel()->init(size);
}

void GameViewPresenterImpl::bind() {
    auto lock = getSemaphore().lock();

    Stage& territory = *miniProgrammingWorld->getStage();
    auto& tilesProperty = territory.tilesProperty();
    
    tilesProperty.addOnAddedListener([this](auto& tile) { addTileNode(tile); });
    tilesProperty.addOnRemovedListener([this](auto& tile) { removeTileNode(tile); });
    tilesProperty.forEach([this](auto& tile) { addTileNode(tile); });

    auto miniProgrammingWorldLog = miniProgrammingWorld->getGameLog();
    miniProgrammingWorldLog->logEntriesProperty().addOnAddedListener([this](const LogEntry& entry) {
        auto lock = getSemaphore().lock();
        addLogEntry(entry);
    });
    miniProgrammingWorldLog->logEntriesProperty().forEach([this](const LogEntry& entry) {
        addLogEntry(entry);
    });

    onBind();
}

void GameViewPresenterImpl::onBind()
{
  // can be overridden by subclasses
}

void GameViewPresenterImpl::playClicked() {
    miniProgrammingWorld->getPerformance()->resumeGame();
}

void GameViewPresenterImpl::pauseClicked() {
    miniProgrammingWorld->getPerformance()->pauseGame();
}

void GameViewPresenterImpl::undoClicked() {
    miniProgrammingWorld->getPerformance()->getGameCommandStack()->undo();
}

void GameViewPresenterImpl::redoClicked() {
    miniProgrammingWorld->getPerformance()->getGameCommandStack()->redo();
}

void GameViewPresenterImpl::close() {
    miniProgrammingWorld->getPerformance()->abortOrStopGame();
}

void GameViewPresenterImpl::textTyped(std::string text) {
    throw std::runtime_error("not implemented");
}

void GameViewPresenterImpl::addTileNode(const mpw::Tile& tile) {
    Tile& nonConstTile = const_cast<Tile&>(tile); // TODO const-correctness: adapt after migrate generator
    const Location& location = nonConstTile.getLocation();

    auto& contentsProperty = nonConstTile.contentsProperty();
    addedContentListenerIds[&tile] = contentsProperty.addOnAddedListener([this, &nonConstTile](const mpw::TileContent&) {
        auto lock = getSemaphore().lock();
        setTileNodeAt(nonConstTile.getLocation(), nonConstTile);
    });
    removedContentListenerIds[&tile] = contentsProperty.addOnRemovedListener([this, &nonConstTile](const mpw::TileContent&) {
        auto lock = getSemaphore().lock();
        setTileNodeAt(nonConstTile.getLocation(), nonConstTile);
    });
    setTileNodeAt(location, tile);
}

void GameViewPresenterImpl::removeTileNode(const mpw::Tile& tile) {
    Tile& nonConstTile = const_cast<Tile&>(tile); // TODO const-correctness: adapt after migrate generator
    auto cell = getViewModel()->getCellAt(
            nonConstTile.getLocation().getRow(),
            nonConstTile.getLocation().getColumn());
    cell->clearLayers();
}

void GameViewPresenterImpl::setTileNodeAt(const Location& location, const Tile& tile) {
    auto cell = getViewModel()->getCellAt(location.getRow(), location.getColumn());
    cell->clearLayers();

    onSetTileNodeAtForCell(*cell, tile);
}

int GameViewPresenterImpl::getRotationForDirection(mpw::Direction direction) {
    switch (direction) {
        case Direction::EAST:
            return 0;
        case Direction::SOUTH:
            return 90;
        case Direction::WEST:
            return 180;
        case Direction::NORTH:
            return 270;
    }
    throw std::runtime_error("Invalid direction!");
}

void GameViewPresenterImpl::addLogEntry(const LogEntry& entry) {
    auto& nonConstEntry = const_cast<LogEntry&>(entry); // TODO const correctness
    auto viewModelEntry = std::make_shared<ViewModelLogEntry>();
    viewModelEntry->setMessage(nonConstEntry.getMessage());
    viewModelEntry->setColor(Color::BLACK);
    getViewModel()->addToLogEntries(viewModelEntry);
}

void GameViewPresenterImpl::speedChanged(double speedValue) {
    if (!(speedValue >= 0 && speedValue <= 10)) {
        throw std::runtime_error("Provided speed is not in range [0, 10]");
    }
    miniProgrammingWorld->getPerformance()->setSpeed(speedValue);
}


}

File: simulators/de.unistuttgart.iste.sqa.mpw.framework.java/core/src/main/java/de/unistuttgart/iste/sqa/mpw/framework/viewmodel/impl/GameViewPresenterImpl.java
package de.unistuttgart.iste.sqa.mpw.framework.viewmodel.impl;

import de.unistuttgart.iste.sqa.mpw.framework.utils.Preconditions;
import de.unistuttgart.iste.sqa.mpw.framework.mpw.*;
import de.unistuttgart.iste.sqa.mpw.framework.viewmodel.Color;
import de.unistuttgart.iste.sqa.mpw.framework.viewmodel.GameViewPresenter;
import de.unistuttgart.iste.sqa.mpw.framework.viewmodel.ViewModelCell;
import de.unistuttgart.iste.sqa.mpw.framework.viewmodel.ViewModelLogEntry;
import javafx.collections.ListChangeListener;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Semaphore;

import static de.unistuttgart.iste.sqa.mpw.framework.utils.Preconditions.checkState;

public abstract class GameViewPresenterImpl extends GameViewPresenter {
	private final MiniProgrammingWorld miniProgrammingWorld;

	private final Map<LogEntry, ViewModelLogEntry> logEntryMap = new HashMap<>();

	public GameViewPresenterImpl(MiniProgrammingWorld miniProgrammingWorld) {
		super(new Semaphore(1, true));
		this.miniProgrammingWorld = miniProgrammingWorld;

		final Size size = this.miniProgrammingWorld.getStage().getStageSize();
		getViewModel().init(size);
	}

	@Override
	public void bind() {
		runLocked(() -> {
			bindTiles();
			bindGameLog();
			bindButtons();
			onBind();
		});
	}

	protected void onBind() {
		// can be overridden by subclasses
	}

	private void bindTiles() {
		var stage = miniProgrammingWorld.getStage();
		var tilesProperty = stage.tilesProperty();
		tilesProperty.addListener(tilesChangedListener);
		tilesProperty.forEach(this::addTileNode);
	}

	private void bindGameLog() {
		var gameLog = miniProgrammingWorld.getGameLog();
		var logEntriesProperty = gameLog.logEntriesProperty();
		logEntriesProperty.addListener(logChangedListener);
		logEntriesProperty.forEach(this::addLogEntry);
	}

	private void bindButtons() {
		var viewModel = getViewModel();
		var modeProperty = miniProgrammingWorld.getPerformance().modeProperty();
		var gameCommandStack = miniProgrammingWorld.getPerformance().getGameCommandStack();
		var anyExecutedCommandsBinding = gameCommandStack.executedCommandsProperty().emptyProperty().not();
		var anyUndoneCommandsBinding = gameCommandStack.undoneCommandsProperty().emptyProperty().not();

		viewModel.playButtonEnabledProperty().bind(modeProperty.isEqualTo(Mode.PAUSED));
		viewModel.pauseButtonEnabledProperty().bind(modeProperty.isEqualTo(Mode.RUNNING));
		viewModel.undoButtonEnabledProperty().bind(anyExecutedCommandsBinding.and(modeProperty.isNotEqualTo(Mode.RUNNING)));
		viewModel.redoButtonEnabledProperty().bind(anyUndoneCommandsBinding.and(modeProperty.isNotEqualTo(Mode.RUNNING)));
	}

	/**
	 * Log Handling
	 */

	private final ListChangeListener<LogEntry> logChangedListener = new ListChangeListener<>() {
		@Override
		public void onChanged(final Change<? extends LogEntry> change) {
			runLocked(() -> {
				while (change.next()) {
					if (change.wasAdded()) {
						change.getAddedSubList().forEach(entry -> addLogEntry(entry));
					}
					if (change.wasRemoved()) {
						change.getRemoved().forEach(entry -> removeLogEntry(entry));
					}
				}
			});
		}
	};

	private void addLogEntry(LogEntry entry) {
		var viewModelLogEntry = new ViewModelLogEntry();
		viewModelLogEntry.setMessage(entry.getMessage());
		viewModelLogEntry.setColor(getColorForLogEntry(entry));
		getViewModel().addToLogEntries(viewModelLogEntry);
		logEntryMap.put(entry, viewModelLogEntry);
	}

	protected Color getColorForLogEntry(LogEntry entry) {
		return Color.BLACK;
	}

	private void removeLogEntry(LogEntry entry) {
		var viewModelLogEntry = logEntryMap.remove(entry);
		getViewModel().removeFromLogEntries(viewModelLogEntry);
	}


	/**
	 * Tile Handling
	 */

	private final ListChangeListener<Tile> tilesChangedListener = new ListChangeListener<Tile>() {
		@Override
		public void onChanged(final Change<? extends Tile> change) {
			runLocked(() -> {
				while (change.next()) {
					if (change.wasAdded()) {
						change.getAddedSubList().forEach(tile -> addTileNode(tile));
					}
					if (change.wasRemoved()) {
						change.getRemoved().forEach(tile -> removeTileNode(tile));
					}
				}
			});
		}
	};

	private void addTileNode(final Tile tile) {
		final Location location = tile.getLocation();
		tile.contentsProperty().addListener((v, c, l) -> {
			runLocked(() -> {
				setTileNodeAt(tile.getLocation(), tile);
			});
		});
		setTileNodeAt(location, tile);
	}

	private void removeTileNode(final Tile tile) {
		var cell = getViewModel().getCellAt(tile.getLocation().getRow(), tile.getLocation().getColumn());
		cell.getLayers().clear();
	}

	private void setTileNodeAt(final Location location, final Tile tile) {
		var cell = getViewModel().getCellAt(location.getRow(), location.getColumn());
		cell.getLayers().clear();
		onSetTileNodeAtForCell(cell, tile);
	}

	protected abstract void onSetTileNodeAtForCell(final ViewModelCell cell, final Tile tile);

	protected int getRotationForDirection(final Direction direction) {
		assert direction != null;

		switch (direction) {
			case EAST:
				return 0;
			case SOUTH:
				return 90;
			case WEST:
				return 180;
			case NORTH:
				return 270;
		}
		throw new RuntimeException("Invalid direction!");
	}

	@Override
	public void playClicked() {
		miniProgrammingWorld.getPerformance().resumeGame();
	}

	@Override
	public void pauseClicked() {
		miniProgrammingWorld.getPerformance().pauseGame();
	}

	@Override
	public void undoClicked() {
		miniProgrammingWorld.getPerformance().getGameCommandStack().undo();
	}

	@Override
	public void redoClicked() {
		miniProgrammingWorld.getPerformance().getGameCommandStack().redo();
	}

	@Override
	public void speedChanged(double speedValue) {
		Preconditions.checkState(speedValue >= 0 && speedValue <= 10,
					"Provided speed is not in range [0, 10]");
		miniProgrammingWorld.getPerformance().setSpeed(speedValue);
	}

	@Override
	public void close() {
		miniProgrammingWorld.getPerformance().abortOrStopGame();
	}

	@Override
	public void textTyped(String text) {
		throw new RuntimeException("not implemented");
	}

	private void runLocked(Runnable runnable) {
		try {
			getSemaphore().acquire();
			runnable.run();
		} catch (final Exception e) {
			throw new RuntimeException(e);
		} finally {
			getSemaphore().release();
		}
	}

}
